<center><font size=5> <b> SMU Homewwork 2 - MSDS7337 - Natural Language Processing </b> </font>
<br><br><br>
<center><font size=4> <b> Spring 2021  Ikenna Nwaogu </b></font> 




<img src="https://cdn.searchenginejournal.com/wp-content/uploads/2020/08/an-introduction-to-natural-language-processing-with-python-for-seos-5f3519eeb8368-760x400.webp"> 
<br>
<p align="center"><font size=4> <b><center> Lexical Diversity, Vocabulary Size and New Text Difficulty Score  </b></font> </p>

In [1]:
import nltk
from nltk.book import *
import numpy as np
from sklearn.preprocessing import minmax_scale

*** Introductory Examples for the NLTK Book ***
Loading text1, ..., text9 and sent1, ..., sent9
Type the name of the text or sentence to view it.
Type: 'texts()' or 'sents()' to list the materials.
text1: Moby Dick by Herman Melville 1851
text2: Sense and Sensibility by Jane Austen 1811
text3: The Book of Genesis
text4: Inaugural Address Corpus
text5: Chat Corpus
text6: Monty Python and the Holy Grail
text7: Wall Street Journal
text8: Personals Corpus
text9: The Man Who Was Thursday by G . K . Chesterton 1908


## 1. Vocabulary size Normalization

Normaliztion can be done using Min Max Scaler from the sklearn module. It can also be done using the simple normalization formula. The formula is dividing the difference of each text's vocabulary size to the min vocabular size on the list by the difference between the max and the min vocabulary size. Texts from nltk package were used.

In [2]:
#### Method to get the Vocab Size and normalize the score

def n_vocab_size(*arg):
    v_size = np.array([])
    v_size_norm = np.array([])
    
    #### Getting the Vocab Size
    for text in arg:
        v_size = np.append(v_size,len(set(text)))
    
    #### Normalizing using formula 
    for vsize in v_size:
        v_size_norm = np.append(v_size_norm,(vsize - v_size.min()) /
                                                    (v_size.max() - v_size.min()))
    
    #### Normalizing using sklearn module
    v_size_norm_sklearn = minmax_scale(v_size, feature_range=(0,1), axis=0)
    
    return(v_size,v_size_norm,v_size_norm_sklearn)



In [3]:
vocab_size = n_vocab_size(text1,text2,text3,text4,text5,text6,text7,text8,text9)

In [4]:
print("Vocabulary Size", *vocab_size[0],sep='\n- ')

Vocabulary Size
- 19317.0
- 6833.0
- 2789.0
- 9913.0
- 6066.0
- 2166.0
- 12408.0
- 1108.0
- 6807.0


In [5]:
print("Normaliztion using the simple formula", *vocab_size[1],sep='\n- ')



Normaliztion using the simple formula
- 1.0
- 0.31440496457795597
- 0.09231698610577187
- 0.4835520896260091
- 0.2722829370091713
- 0.05810313581196112
- 0.6205722444944808
- 0.0
- 0.31297709923664124


In [6]:
print("Normalization using the sklearn module", *vocab_size[2],sep='\n- ')

Normalization using the sklearn module
- 0.9999999999999999
- 0.3144049645779559
- 0.09231698610577188
- 0.48355208962600904
- 0.2722829370091713
- 0.05810313581196112
- 0.6205722444944807
- 0.0
- 0.3129770992366412


## 2. Long Word Vocabulary Size and Normalization With Default Length 10.

Below we used the same normalization methods as we used with the overall vocabulary size. Using the texts from nltk package we computed the long word vocabulary size. The function written below computes the values and normalization values of the long word vocabulary sizes of multiple texts.

In [7]:
def n_long_vocab_size(*arg, longWordLength = 10):
    v_size = np.array([])
    v_size_norm = np.array([])
    
    #### Getting the Vocab Size
    for text in arg:
        V = set(text)
        long_words = [w for w in V if len(w) > longWordLength]
        v_size = np.append(v_size,len(set(long_words)))
    
    #### Normalizing using formula 
    for vsize in v_size:
        v_size_norm = np.append(v_size_norm,(vsize - v_size.min()) /
                                                    (v_size.max() - v_size.min()))
    
    #### Normalizing using sklearn module
    v_size_norm_sklearn = minmax_scale(v_size, feature_range=(0,1), axis=0)
    
    return(v_size,v_size_norm,v_size_norm_sklearn)

In [8]:
long_vocab_size = n_long_vocab_size(text1,text2,text3,text4,text5,text6,text7,text8,text9,longWordLength = 10)

In [9]:
print("Vocabulary Size", *long_vocab_size[0],sep='\n- ')

Vocabulary Size
- 1912.0
- 724.0
- 64.0
- 1378.0
- 259.0
- 45.0
- 1463.0
- 38.0
- 478.0


In [10]:
print("Normaliztion using the simple formula", *long_vocab_size[1],sep='\n- ')

Normaliztion using the simple formula
- 1.0
- 0.3660618996798292
- 0.013874066168623266
- 0.7150480256136607
- 0.11792956243329776
- 0.003735325506937033
- 0.7604055496264674
- 0.0
- 0.23479188900747064


In [11]:
print("Normalization using the sklearn module", *long_vocab_size[2],sep='\n- ')

Normalization using the sklearn module
- 1.0000000000000002
- 0.3660618996798293
- 0.013874066168623266
- 0.7150480256136607
- 0.11792956243329776
- 0.0037353255069370317
- 0.7604055496264674
- 0.0
- 0.23479188900747067


## 3. Text Difficulty Score

In this section a new scoring metrics were used. It is made up by combining the lexical score with the normalized score of the vocabulary size and the long-word vocabulary size. 

In [29]:
def lexical_diversity(*args):
    lex_div = np.array([])
    for text in args:
        val = len(set(text)) / len(text)
        lex_div = np.append(val)
    return(lex_div)

def text_difficulty_score(*args, longWordLength = 10):
    lex_div = np.array([])
    longword_voc = np.array([])
    voc_size = np.array([])
    v_size_norm = np.array([])
    df_scores_norm = np.array([])
    longword_voc_norm = np.array([])
    
    for text in args:
        V = set(text)
        val = len(set(text)) / len(text)
        lex_div = np.append(lex_div,val)
        long_words = [w for w in V if len(w) > longWordLength]
        longword_voc = np.append(longword_voc,len(set(long_words)))
        voc_size = np.append(voc_size,len(set(text)))
        
    # Normalize the score of vocabulary size
    for vsize in voc_size:
        v_size_norm = np.append(v_size_norm,(vsize - voc_size.min()) /
                                                    (voc_size.max() - voc_size.min()))
    
    # Normalize the long word score of vocabulary size
    for vsize in longword_voc:
        longword_voc_norm = np.append(longword_voc_norm,(vsize - longword_voc.min()) /
                                                    (longword_voc.max() - longword_voc.min()))
    df_scores = lex_div + longword_voc + v_size_norm
    df_scores1 = lex_div + longword_voc_norm + v_size_norm
    
    ### Normalizing using formula 
    for df_score in df_scores:
        df_scores_norm = np.append(df_scores_norm,(df_score - df_scores.min()) /
                                                    (df_scores.max() - df_scores.min()))
    return(df_scores,df_scores1,df_scores_norm)
#     return(lex_div, longword_voc , voc_size)

###### Dramatic Reader for Lower Grades by Florence Holbrook

In [13]:
import nltk
f=open('/home/ikenna/Documents/datascience/DS7337_NLP/27764-8.txt','r',encoding="ISO-8859-1")
raw=f.read()
tokens = nltk.word_tokenize(raw)
text = nltk.Text(tokens)

###### Graded Poetry: Seventh Year by Georgia Alexander and Katherine Devereux Blake

In [14]:
f=open('/home/ikenna/Documents/datascience/DS7337_NLP/9542-8.txt','r',encoding="ISO-8859-1")
raw=f.read()
tokens = nltk.word_tokenize(raw)
text1 = nltk.Text(tokens)

###### Sanders' Union Fourth Reader by Charles W. Sanders

In [15]:
f=open('/home/ikenna/Documents/datascience/DS7337_NLP/9542-8.txt','r',encoding="ISO-8859-1")
raw=f.read()
tokens = nltk.word_tokenize(raw)
text2 = nltk.Text(tokens)

In [30]:
score = text_difficulty_score(text,text1, text2)

In [31]:
print("New Difficulty Score with long-word vocabulary size normalized", *score[1],sep='\n- ')

New Difficulty Score with long-word vocabulary size normalized
- 0.12242019074488639
- 2.211183114238104
- 2.211183114238104


In [32]:
print("New Difficulty Score without long-word vocabulary size normailized", *score[0],sep='\n- ')

New Difficulty Score without long-word vocabulary size normailized
- 161.1224201907449
- 257.2111831142381
- 257.2111831142381


In [33]:
print("New Normalized Difficulty Score", *score[2],sep='\n- ')

New Normalized Difficulty Score
- 0.0
- 1.0
- 1.0


The vocabulary size of all 3 text are 3992, 5703 and 17105; the lexical diversity is 0.1224, 0.2111 and 0.1186 for Text 1, Text 2 and Text 3 respectively. 

The difficulty scores are 161, 257, and 257 for Text 1, Text 2 and Text 3 respectively. Comparing this score to both lexical and vocabulary size, they are have diverent score rankings. I initially thought that any other scores could be closely related to lexical diversity score. The findings shows how long-word vocabulary size affects the difficulty score. The length of the word used for the long-word vocabulary size can be critical on the difficullty of the text.

Overall the long-word vocabulary size is important and a good way to measure text difficulty. Lexical diversity is better because it takes into account the length of the text. For vocabulary size, if a text is large then the vocabulary size is most likely going to be large which would not be good. I think combining the 3 is better given that the vocabulary size would be normalized.